docs: Adding Spark Best Practices (#833)
awslabs/data-on-eks: website/docs/bestpractices/analytics/spark-best-practices.md

---
sidebar_position: 3
sidebar_label: Spark on EKS Best Practices
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CollapsibleContent from '../../../src/components/CollapsibleContent';

# Spark on EKS Best Practices

This page aims to provide comprehensive best practices and guidelines for deploying, managing, and optimizing Apache Spark workloads on Amazon Elastic Kubernetes Service (EKS). This helps organizations to successfully run and scale their Spark Applications at scale in a containerised environment on Amazon EKS.

For deploying Spark on EKS you can leverage the [blueprints](https://awslabs.github.io/data-on-eks/docs/blueprints/data-analytics/spark-operator-yunikorn), which readily incorporates most of the best practices. You can further customize this blueprint, to tweak the configurations to match your specific application requirements and environment constraints, as outlined in this guide.

<CollapsibleContent header={<h2><span>Networking Best Practices</span></h2>}>

> ## EKS Networking

Reviewer: We have Ipv6 blueprints available in this repo. Worth mentioning?

    * It is also worth noting that both worker nodes and pods require IP addresses. By default, VPC CNI has WARM_ENI_TARGET=1 means that ipamd should keep "a full ENI" of available IPs around in the IPAMD warm pool for the Pod IP assignment. 
    * For new EKS clusters, it is recommended to over-provision the subnets you will use for Pod networking for growth.
* Remediation for IP Address exhaustion
    * Ideally, we recommend you to over-provision your subnets to accommodate future growth. Although, IP exhaustion remediation methods exist for VPCs, they introduce additional operational complexity and have consequential implications to consider.
    * One of the first options to consider for addressing IP address exhaustion, is creating new subnets with larger CIDR blocks within the VPC and deploying worker nodes in these expanded subnets.
    * If adding more subnets, is not an option, then you will have to work on optimising the IP address assignment by tweaking CNI Configuration Variables as described below. Also evaluate the implications of tweaking these variables.
    * VPC CNI Variables → WARM_IP_TARGET & MINIMUM_IP_TARGET

>         * The default behaviour of `WARM_ENI_TARGET=1` can be changed by configuring  `MINIMUM_IP_TARGET` and `WARM_IP_TARGET` variables. This configuration, optimizes the number of WARM IPs held at Instance level. The VPC CNI's ipamd regularly communicates with the EC2 API to maintain and monitor a WARM IPs.

Author: Done

        * The appropriate values for these settings will vary based on your specific instance type and anticipated pod density.
        * It is recommended to set MINIMUM_IP_TARGET to slightly higher number than the `maximum` expected number of pods (driver/executors) that you plan to run on each node. This reduces the frequency of EC2 API calls required for managing ENI and IP address assignments.
        * Implication:
            * Careful consideration is needed with this configuration as it increases EC2 API call frequency for IP attachment and detachment operations by ipamd. Excessive API calls can trigger throttling, preventing new ENI or IP assignments across the entire cluster.
            * Hence this configuration is ideally recommended for small clusters or clusters with very low pod churn. 
### CoreDNS Recommendations
    * VPC resolver has a hard cap of 1024 packets per second for each ENI. By default, an EKS cluster runs two replicas of CoreDNS. Anything beyond 1024 packets per second will be throttled, and will result in `unknownHostException`. 

>     * While running Spark applications on Kubernetes, there could be scenarios where executors might perform significant number of external DNS lookups. It could be that Spark application, often needs to communicate with external services or databases, for data ingestion, processing, or need to reach external shuffle service. 

Reviewer: Let's make this more concise. consolidate "could", "might"

    * To address the scalability of default coreDNS, consider implementing one of the following two options:
        * Enable coreDNS auto-scaling.

>         * Implement Node local cache.

Reviewer: Let's provide links.

        * Reduce inter-AZ traffic costs
        * Reduce network latency between executors/Pods
    * To have pods co-located on the same AZ, we can use podAffinity based scheduling constraints. The scheduling constraint preferredDuringSchedulingIgnoredDuringExecution can be enforced in the Pod spec. For example, in Spark we can use a custom template for our driver and executor pods:

```
spec:
  executor:
    affinity:
      podAffinity:
        preferredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
              matchExpressions:
              - key: sparkoperator.k8s.io/app-name
                operator: In
                values:
                - <<spark-app-name>>
          topologyKey: topology.kubernetes.io/zone
          ...
```

</CollapsibleContent>

<CollapsibleContent header={<h2><span>Scaling Best Practices</span></h2>}>

> ## Karpenter Recommendations

Reviewer: Let's link to Karpenter site.

>     ### NodePools configuration

Reviewer: Fix indentation please.

        Consider creating separate NodePools for driver and executor pods. 

>         * Driver Nodepool

Reviewer: Driver Nodepool and executor Nodepool sections should be under their own headers.

>             * Spark driver will be a single replica, and manages the entire lifecycle of the Spark application. Terminating spark driver pod, effectively means terminating the entire spark job. Hence: 

Reviewer: ```suggestion
The Spark driver is a single pod and manages the entire lifecycle of the Spark application. Terminating spark driver pod, effectively means terminating the entire spark job.
```

>                 * Configure Driver Nodepool to use on-demand nodes only. 

Reviewer: Why is this recommended? Not clear to me.

            * This helps in dynamic environment by reducing the need to recompute the shuffle data or RDD blocks, there by improving the job completion times and resource efficiency.

</CollapsibleContent>

<CollapsibleContent header={<h2><span>Advanced Scheduling</span></h2>}>
## Scheduling Considerations
### Default Kubernetes Scheduler behaviour.
    * Default kubernetes scheduler uses `least allocated` approach. This strategy aims to distribute pods evenly across cluster, which helps in maintaining availability and a balanced resource utilization across all nodes, rather than packing more pods in fewer nodes.
    * `Most allocated` approach on the other hand, aims to favor nodes with most amount of allocated resources, which leads to packing more pods onto nodes that are already heavily allocated. This approach is favourable for spark jobs, as it aims for high utilization on select nodes at pod scheduling time, leading to better consolidation of nodes. You will have to leverage a custom kube-scheduler with this option enabled, or leverage Custom Schedulers purpose built for more advanced orchestration.
### Custom Schedulers

>     * Custom schedulers enhance Kubernetes’ native scheduling capabilities by providing advanced features tailored for batch and high-performance computing workloads. Custom schedulers enhance resource allocation by optimizing bin-packing and offering scheduling tailored to specific application needs. Here are most frequently used custom schedulers for running Spark workloads on Kubernetes.

Reviewer: ```suggestion
Custom schedulers enhance Kubernetes’ native scheduling capabilities by providing advanced features tailored for batch and high-performance computing workloads. Custom schedulers enhance resource allocation by optimizing bin-packing and offering scheduling tailored to specific application needs. Here are popular custom schedulers for running Spark workloads on Kubernetes.
```

        * [Apache Yunikorn](https://yunikorn.apache.org/)
        * [Volcano](https://volcano.sh/en/)
    * Advantages of leveraging custom schedulers like Yunikorn.
        * It provides hierarchical queue system and configurable policies allow for complex resource management.
        * Gang Scheduling
    * How will Yunikorn and Karpenter work together?
        * Karpenter focuses on node provisioning and scaling, while custom schedulers like Yunikorn, brings in application awareness for scheduling.
        * Yunikorn manages the advanced application scheduling aspects like queue management, resource fairness, gang scheduling. 
        * Once Yunikorn schedules the pods, based on application awareness, Karpenter will then fulfil needs of pending pods, spawned by Yunikorn.

</CollapsibleContent>

<CollapsibleContent header={<h2><span>Storage Best Practices</span></h2>}>
## Storage Considerations

> ### Node Storage

Reviewer: Worth mentioning local nvme storage?
